### PR TITLE
Handle old events

### LINKED ISSUE #6333: Properly handle old events in handleConsensusRound
Once Platform team has implemented the pre-requisite for [HIP-721](https://github.com/hashgraph/hedera-improvement-proposal/blob/main/HIP/hip-721.md): [Add software version to events](https://github.com/hashgraph/hedera-services/issues/6026), we need to properly handle old events in `handleConsensusRound`.

By looking at the new software version contained in the events, when Services processes an event, it should determine if the event was generated by a previous version of the software, and if so, the transaction should fail and the user should get an error along the lines of "Please Retry" when they fetch the receipt for the transaction.

This needs to be done for mono-service and non-mono-service workflows.

### PR DESCRIPTION
**Pull request by Neeharika-Sompalli**
Fixes #6333 

- Added a check to validate the software version of the event in `handleConsensusRound`
- If the version is older add a `BUSY` status to the TransactionReceipt and store in RecordCache

Testing:

- [x] Unit Tests
- [ ] End to End Tests

### CODE REVIEW

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/ProcessLogic.java [hunk 1]
```diff
@@ -35,15 +36,17 @@ public interface ProcessLogic {
...
     /**
-     * Orchestrates a process to express the full implications of the given consensus transaction at
-     * the specified time.
+     * Orchestrates a process to express the full implications of the given consensus transaction at the specified
+     * time.
      *
-     * @param platformTxn the consensus transaction to incorporate.
+     * @param platformTxn      the consensus transaction to incorporate.
      * @param submittingMember the id of the member that submitted the txn
+     * @param softwareVersion
```
Comment (mhess-swl) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/ProcessLogic.java:48]: Note here would be helpful

**Review by mhess-swl — APPROVED** [review]

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java [hunk 2]
```diff
@@ -81,12 +87,28 @@ public StandardProcessLogic(
...
+    public void incorporateConsensusTxn(
+            @NonNull final ConsensusTransaction platformTxn,
+            final long submittingMember,
+            @NonNull final SoftwareVersion softwareVersion) {
+
         try {
             final var accessor = expandHandleSpan.accessorFor(platformTxn);
+            // Check if the transaction is from an older event version.
+            // If so, return and set the status on the receipt to BUSY
+            if (!SEMANTIC_VERSIONS.deployedSoftwareVersion().equals(softwareVersion)) {
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java:103]: Other scenario, config only software version bump.  Other node has it, this node doesn't.

**Review by edward-swirldslabs — APPROVED** [review]

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java [hunk 3]
```diff
@@ -81,12 +87,28 @@ public StandardProcessLogic(
...
+                log.info(
```
Comment (tinker-michaelj) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java:104]: Imo we don't want to log at `info` in a non-exceptional `handle()` code path that could feasibly happen thousands of times.

**Review by tinker-michaelj — DISMISSED** [review]
One comment on `log.info()`, otherwise this all makes sense to me!

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/ProcessLogic.java (hunk 1 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/ProcessLogic.java:48]: yes missed it. Fixed. Thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java:103]: I wasn't sure with current upgrades if it is even possible for a newer version to submit transactions to older version.  I thought it is irrelevant.

**Review by mhess-swl — DISMISSED** [review]

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (nickpoorman) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java:103]: At the moment, I don't think it's possible for an older version of the protocol even to establish a connection to a newer version given the version that's exchanged. However, I would assume this is probably more in line with what we want to achieve here.

```
SEMANTIC_VERSIONS.deployedSoftwareVersion().compareTo(softwareVersion) > 0
```

[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java:103]: Thanks @nickpoorman . 

Did the same check previously. But on review, `compareTo` was creating a regex matcher in `alphaNumberOf` [here](https://github.com/hashgraph/hedera-services/blob/6333-discard-old-events/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/context/properties/SerializableSemVers.java#L44). So @tinker-michaelj  suggested we do a simple equals comparison. 

@tinker-michaelj  Is it okay to do this check directly using `compareTo` or do you have any other suggestions?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java (hunk 3 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java]: Fixed. Thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/records/RecordCache.java [hunk 4]
```diff
@@ -76,12 +77,46 @@ public void setFailInvalid(
...
+    /**
+     * Set the record for a transaction that was submitted with an older event version. We set the status of the
+     * transaction as {@code ResponseCodeEnum.BUSY} on the receipt. This transaction needs to be re-submitted for
+     * it to succeed.
+     * @param effectivePayer payer for the transaction
+     * @param accessor transaction accessor
+     * @param consensusTimestamp consensus timestamp
+     * @param submittingMember submitting member
+     */
+    public void setStaleTransaction(
```
Comment (hendrikebbers) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/records/RecordCache.java:92]: null annotations missing

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/records/RecordCache.java [hunk 5]
```diff
@@ -76,12 +77,46 @@ public void setFailInvalid(
...
+
+    /**
+     * Create a failure record with the given status in the receipt and store it in cache.
+     * @param effectivePayer payer for the transaction
+     * @param accessor transaction accessor
+     * @param consensusTimestamp consensus timestamp
+     * @param submittingMember submitting member
+     * @param status status of the transaction
+     */
+    private void setRecordWithStatus(
```
Comment (hendrikebbers) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/records/RecordCache.java:108]: null annotations missing

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java [hunk 6]
```diff
@@ -68,7 +73,8 @@ public StandardProcessLogic(
             final ScheduleProcessing scheduleProcessing,
             final ExecutionTimeTracker executionTimeTracker,
             final RecordStreaming recordStreaming,
-            final StateView workingView) {
+            final StateView workingView,
+            final RecordCache recordCache) {
```
Comment (hendrikebbers) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java:77]: null annotations missing

Comment (Neeharika-Sompalli) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java:77]: Added . Thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/records/RecordCache.java (hunk 5 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/records/RecordCache.java:108]: Added. Thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/records/RecordCache.java (hunk 4 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/records/RecordCache.java:92]: Added . Thanks !

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/StandardProcessLogic.java]: Changed to use `isAfter` . Added the alpha number comparisons to use regex pattern only when creating new object.

**Review by edward-swirldslabs — DISMISSED** [review]

**Review by edward-swirldslabs — APPROVED** [review]

**Review by tinker-michaelj — APPROVED** [review]
All makes sense to me!

**Review by mhess-swl — DISMISSED** [review]

**Review by tinker-michaelj — APPROVED** [review]
LGTM

**Pull request merged**